### PR TITLE
Add a visualizer around the OsuLogo

### PR DESCRIPTION
**Pull request by ColdVolcano**
- [ ] Requieres https://github.com/ppy/osu-framework/pull/701

Note this won't work with the osu bgm

### CODE REVIEW

File: osu.Game/Screens/Menu/MenuVisualisation.cs [hunk 1]
```diff
@@ -1,11 +1,128 @@
...
+                    ControlPoint kiaiControlPoint;
+                    beatmap.Value.Beatmap.TimingInfo.TimingPointAt(beatmap.Value.Track.CurrentTime, out kiaiControlPoint);
+                    kiai.Value = (kiaiControlPoint?.KiaiMode ?? false);
+
+
+                    if (timeOfLastUpdate + 50 <= Clock.CurrentTime)
+                    {
+                        int i = 0;
+
+                        audioData = (beatmap.Value.Track as TrackBass).GetChannelData();
```
Comment (Tom94) [inline osu.Game/Screens/Menu/MenuVisualisation.cs]: Raw bass methods should never be called from the main thread. Probably should make `GetChannelData` an internal thing to the framework, store its result in a local variable, and copy that local variable whenever it's requested from here. Similar to how other bass calls are used to update values inside the `Update` method of the audio tracks (which is always invoked from the audio thread).

That aside, your code here will crash if `Track` is not `TrackBass` which is not okay. It should be implementation agnostic.

File: osu.Game/Screens/Menu/OsuLogo.cs [hunk 2]
```diff
@@ -33,6 +33,8 @@ public class OsuLogo : Container
 
         private readonly Container colourAndTriangles;
 
+        private readonly MenuVisualisation visualizer;
```
Comment (peppy) [inline osu.Game/Screens/Menu/OsuLogo.cs]: why is it called menu visualisation when it's inside the osu logo

File: osu.Game/Screens/Menu/MenuVisualisation.cs [hunk 3]
```diff
@@ -1,11 +1,129 @@
...
+                    }
+                }
+            }
+        }
+
+        protected override void UpdateAfterChildren()
+        {
+            foreach (Box b in Children)
+            {
+                b.ScaleTo(new Vector2(1, b.Scale.Y > 0.25f ? b.Scale.Y - 0.1f : b.Scale.Y * 0.8f), 50);
```
Comment (peppy) [inline osu.Game/Screens/Menu/MenuVisualisation.cs]: magic numbers everywhere

File: osu.Game/Screens/Menu/MenuVisualisation.cs [hunk 4]
```diff
@@ -1,11 +1,129 @@
...
+
+                if (b.Scale.Y > 0)
+                {
+                    b.Alpha = MathHelper.Clamp(b.Scale.Y - 0.01f, 0, 0.15f) * .2f / 0.15f;
```
Comment (peppy) [inline osu.Game/Screens/Menu/MenuVisualisation.cs]: magic numbers everywhere

File: osu.Game/Screens/Menu/MenuVisualisation.cs [hunk 5]
```diff
@@ -1,11 +1,129 @@
...
+                            if (audioData[index] * 3.8f >= b.Scale.Y)
+                            {
+                                b.ClearTransforms(true);
+                                b.Scale = new Vector2(1, audioData[index] * 3.8f);
+                            }
+                            i++;
+                        }
+
+                        indexOffset += 5;
+                        if (indexOffset > 250)
```
Comment (peppy) [inline osu.Game/Screens/Menu/MenuVisualisation.cs]: magic numbers everywhere

File: osu.Game/Screens/Menu/MenuVisualisation.cs [hunk 6]
```diff
@@ -1,11 +1,129 @@
...
+                    {
+                        int i = 0;
+
+                        audioData = beatmap.Value.Track.FrequencyAmplitudes;
+
+                        foreach (Box b in Children)
+                        {
+                            int index = (i % bars_per_visualizer) + indexOffset;
+                            if (index > 249)
+                                index -= 250;
```
Comment (peppy) [inline osu.Game/Screens/Menu/MenuVisualisation.cs]: magic numbers everywhere

File: osu.Game/Screens/Menu/MenuVisualisation.cs [hunk 7]
```diff
@@ -1,11 +1,129 @@
...
+                    {
+                        Anchor = Anchor.BottomCentre,
+                        Origin = Anchor.BottomCentre,
+                        Size = new Vector2(barWidth, 300),
+                        RelativePositionAxes = Axes.Both,
+                        Scale = new Vector2(1, 0),
+                        Colour = Color4.White,
+                        Alpha = 0.2f,
+                        Position = new Vector2(
+                        -(float)Math.Sin((float)(j + i * (bars_per_visualizer / 5)) / bars_per_visualizer * 2 * MathHelper.Pi) / 2,
```
Comment (peppy) [inline osu.Game/Screens/Menu/MenuVisualisation.cs]: magic numbers everywhere
repeated code everywhere

File: osu.Game/Screens/Menu/MenuVisualisation.cs [hunk 8]
```diff
@@ -1,11 +1,129 @@
...
+
+        private int indexOffset = 0;
+        private double timeOfLastUpdate = double.MinValue;
+        private float[] audioData;
+
+        private const int bars_per_visualizer = 250;
+
+        public MenuVisualisation()
+        {
+            audioData = new float[256];
```
Comment (peppy) [inline osu.Game/Screens/Menu/MenuVisualisation.cs]: magic numbers everywhere

File: osu.Game/Screens/Menu/MenuVisualisation.cs [hunk 9]
```diff
@@ -1,11 +1,129 @@
...
+        protected override void Update()
+        {
+            if (beatmap?.Value != null)
+            {
+                if ((bool)beatmap.Value?.Track?.IsRunning)
+                {
+                    ControlPoint kiaiControlPoint;
+                    beatmap.Value.Beatmap.TimingInfo.TimingPointAt(beatmap.Value.Track.CurrentTime, out kiaiControlPoint);
+                    kiai.Value = (kiaiControlPoint?.KiaiMode ?? false);
+
```
Comment (peppy) [inline osu.Game/Screens/Menu/MenuVisualisation.cs]: excess newlines

File: osu.Game/Screens/Menu/MenuVisualisation.cs [hunk 10]
```diff
@@ -1,11 +1,129 @@
...
+
+                    if (timeOfLastUpdate + 50 <= Clock.CurrentTime)
```
Comment (peppy) [inline osu.Game/Screens/Menu/MenuVisualisation.cs]: magic numbers everywhere

File: osu.Game/Screens/Menu/MenuVisualisation.cs [hunk 11]
```diff
@@ -1,11 +1,129 @@
...
+                        foreach (Box b in Children)
+                        {
+                            int index = (i % bars_per_visualizer) + indexOffset;
+                            if (index > 249)
+                                index -= 250;
+
+                            if (audioData[index] * 3.8f >= b.Scale.Y)
+                            {
+                                b.ClearTransforms(true);
+                                b.Scale = new Vector2(1, audioData[index] * 3.8f);
```
Comment (peppy) [inline osu.Game/Screens/Menu/MenuVisualisation.cs]: magic numbers everywhere

File: osu.Game/Screens/Menu/MenuVisualisation.cs [hunk 12]
```diff
@@ -1,11 +1,129 @@
...
+
+                        audioData = beatmap.Value.Track.FrequencyAmplitudes;
+
+                        foreach (Box b in Children)
+                        {
+                            int index = (i % bars_per_visualizer) + indexOffset;
+                            if (index > 249)
+                                index -= 250;
+
+                            if (audioData[index] * 3.8f >= b.Scale.Y)
```
Comment (peppy) [inline osu.Game/Screens/Menu/MenuVisualisation.cs]: magic numbers everywhere

File: osu.Game/Screens/Menu/MenuVisualisation.cs [hunk 13]
```diff
@@ -1,11 +1,129 @@
...
 {
-    internal class MenuVisualisation : Drawable
+    internal class MenuVisualisation : Container
     {
+        public override bool HandleInput => false;
+
+        private Bindable<WorkingBeatmap> beatmap;
+        private Bindable<bool> kiai = new Bindable<bool>();
+
+        private int indexOffset = 0;
```
Comment (peppy) [inline osu.Game/Screens/Menu/MenuVisualisation.cs]: `= 0` not needed

File: osu.Game/Screens/Menu/MenuVisualisation.cs [hunk 14]
```diff
@@ -1,11 +1,129 @@
...
+        private double timeOfLastUpdate = double.MinValue;
+        private float[] audioData;
```
Comment (peppy) [inline osu.Game/Screens/Menu/MenuVisualisation.cs]: why even store this when you're getting it every update

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Menu/MenuVisualisation.cs [hunk 15]
```diff
@@ -1,11 +1,129 @@
...
+                    });
+                }
+            }
+
+            kiai.ValueChanged += updateKiai;
+        }
+
+        private void updateKiai(bool newValue)
+        {
+            FadeTo(newValue ? 2.5f : 1, 75);
```
Comment (peppy) [inline osu.Game/Screens/Menu/MenuVisualisation.cs]: fading to 250% alpha?

**Comment by peppy** [comment]
Thanks for your contribution. Unfortunately your code is not up to the standards of this project. Please use the above feedback as a general direction of where you should be going , but review *yourself* every line of your code and try and improve before submitting it for us to consider.

**Comment by ColdVolcano** [comment]
Will close this for the time being.

**Pull request closed without merging**